libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/NavigationOkHttpService.kt

```diff
@@ -0,0 +1,145 @@
+package com.mapbox.navigation.navigator.internal
+
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.bindgen.Expected
+import com.mapbox.bindgen.ExpectedFactory
+import com.mapbox.common.HttpRequest
+import com.mapbox.common.HttpRequestError
+import com.mapbox.common.HttpRequestErrorType
+import com.mapbox.common.HttpResponse
+import com.mapbox.common.HttpResponseCallback
+import com.mapbox.common.HttpResponseData
+import com.mapbox.common.HttpServiceInterface
+import com.mapbox.navigation.navigator.BuildConfig
+import java.io.IOException
+import java.util.HashMap
+import java.util.Locale
+import java.util.UUID
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.logging.HttpLoggingInterceptor
+
+internal class NavigationOkHttpService(
+    private val logger: Logger? = null
+) : HttpServiceInterface() {
+
+    companion object {
+        private const val HEADER_ENCODING = "Accept-Encoding"
+        private const val GZIP = "gzip, deflate"
+    }
+
+    private val requestMap = hashMapOf<Long, Call>()
+
+    private val httpClient: OkHttpClient by lazy {
+        val clientBuilder: OkHttpClient.Builder = OkHttpClient.Builder()
+        if (BuildConfig.DEBUG) {
+            val interceptor = HttpLoggingInterceptor(
+                HttpLoggingInterceptor.Logger { message -> logger?.d(msg = Message(message)) }
+            ).setLevel(HttpLoggingInterceptor.Level.BASIC)
+
+            clientBuilder.addInterceptor(interceptor)
+        }
+        clientBuilder.build()
+    }
+
+    override fun setMaxRequestsPerHost(max: Byte) {
+        httpClient.dispatcher().maxRequestsPerHost = max.toInt()
+    }
+
+    override fun supportsKeepCompression(): Boolean {
+        return true
+    }
+
+    override fun request(request: HttpRequest, callback: HttpResponseCallback): Long {
+        val requestBuilder = Request.Builder()
+        val resourceUrl: String = request.url
+        requestBuilder.url(resourceUrl).tag(resourceUrl.toLowerCase(Locale.US))
+        for ((key, value) in request.headers) {
+            requestBuilder.addHeader(key!!, value!!)
+        }
+        if (request.keepCompression) {
+            // Adding this header manually means okhttp will not automatically decompress the data
+            requestBuilder.addHeader(
+                HEADER_ENCODING,
+                GZIP
+            )
+        }
+        val call = httpClient.newCall(requestBuilder.build())
+        call.enqueue(
+            HttpCallback(
+                logger,
+                request,
+                callback
+            )
+        )
+
+        var id = UUID.randomUUID().mostSignificantBits and Long.MAX_VALUE
+        while (requestMap.containsKey(id)) {
```

Comment:
what's the while loop for? seems like we could store the UUID in the map, it has an equals

Reply:
If there's a random `UUID` generated that we've already generated in the past, I don't want to override it and lose the reference to the call (in case it needs to be canceled). This mechanism isn't implemented on the NN side yet though /cc @vmihaylenko.

Reply:
don't think you'll ever generate the same UUID, ever

["50% probability of at least one collision is 2.71 quintillion"](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions)

Reply:
Huh, yes, this is definitely over-caution. The probability might increase because we use the bitmask though, but still, rather unlikely.